src/main/java/picard/util/AlleleSubsettingUtils.java

```
@@ -0,0 +1,260 @@
+
+        final byte refAllele = referenceAlleleMaybe.get();
+        final byte otherAllele = snp.getAllele1() == refAllele ? snp.getAllele2() : snp.getAllele1();
+
+        // do we have otherAllele in ctx?
+        final Optional<Allele> altAlleleMaybe = ctx.getAlternateAlleles()
+                .stream()
+                .filter(a -> a.length() == 1 && a.getBases()[0] == otherAllele)
+                .findAny();
+
+        if (altAlleleMaybe.isPresent()) {
+            if (ctx.getAlleles().size() == 2) return ctx;
+
+            return AlleleSubsettingUtils.subsetAlleles(ctx, Arrays.asList(ctx.getReference(), altAlleleMaybe.get()));
+        }
+
+        // if not, perhaps we have NON_REF_ALLELE
+        final Optional<Allele> nonRefAlleleMaybe = ctx.getAlternateAlleles().stream().filter(a -> a.equals(NON_REF_ALLELE)).findAny();
+        if (nonRefAlleleMaybe.isPresent()) {
+
+            final VariantContext vcSubsetted = ctx.getAlleles().size() == 2 ? ctx : AlleleSubsettingUtils.subsetAlleles(ctx, Arrays.asList(ctx.getReference(), nonRefAlleleMaybe.get()));
+            return AlleleSubsettingUtils.swapAlleles(vcSubsetted, NON_REF_ALLELE, Allele.create(otherAllele));
+        }
+        return null;
+    }
+
+    public static VariantContext subsetAlleles(final VariantContext originalVc, final List<Allele> allelesToKeep) {
+        VariantContextBuilder vcBuilder = new VariantContextBuilder(originalVc).alleles(allelesToKeep);
+
+        GenotypesContext newGenotypes = subsetAlleles(originalVc.getGenotypes(), originalVc.getAlleles(), allelesToKeep);
+        vcBuilder.genotypes(newGenotypes);
+        return vcBuilder.make();
+    }
+
+    /**
+     * swaps one of the alleles in a VC (and its genotypes) with another.
+     */
+    public static VariantContext swapAlleles(final VariantContext originalVc, final Allele oldAllele, final Allele newAllele) {
+        if (!originalVc.getAlleles().contains(oldAllele)) return originalVc;
```

This seems like it should throw instead of returning.  I'd expect to get back a a variant context that includes the allele I was swapping in.  If you don't think it should change behavior, then expand the doc to mention this case.

---

fixed. and added unit tests.